### PR TITLE
fix(perf): remove @st.cache_resource from train_model and add load_pipeline

### LINKED ISSUE #99: [Bug] train_model() sklearn pipeline is trained and cached at every cold start but never used for prediction — wastes 20-30s startup time with dead code
## Description

`application.py` calls `train_model()` at module load time, which reads both CSV files, engineers all features, and fits a full XGBoost pipeline inside a `@st.cache_resource` block:

```python
# application.py line ~1076
@st.cache_resource
def train_model():
    matches = pd.read_csv("matches.csv")
    deliveries = pd.read_csv("deliveries.csv")
    # ... full feature engineering ...
    pipe = Pipeline([
        ('preprocessor', preprocessor),
        ('model', XGBClassifier(n_estimators=100, ...))
    ])
    pipe.fit(X, y)
    return pipe

pipe = train_model()  # Called at startup — takes ~20-30s on cold start
```

However, `pipe` is **never referenced again** in the entire file. When the user clicks **Run Analysis**, the prediction code bypasses `pipe` entirely and loads a completely different model from disk:

```python
# application.py — inside the `if analyze:` block
loaded_xgb = xgb.XGBClassifier()
loaded_xgb.load_model('xgb_win_prob_model.json')  # Loads separate file every click
```

The two models are also incompatible — `train_model()` uses sklearn feature names (`target`, `wickets`) while the prediction manually builds a feature dict with different names (`target_score`, `wickets_left`). Passing `pipe.predict()` the prediction input would raise a feature name mismatch error.

## Impact

1. **Startup performance**: Every Streamlit cold start (server restart, Streamlit Cloud deployment, sleep-wake cycle) triggers a full model train on ~750k delivery rows. This adds 20-30 seconds of blocking startup time that provides zero benefit to the user.
2. **Dead code**: `pipe` is assigned but never read. Any future contributor who attempts to use `pipe` for predictions will get incorrect results due to the feature name mismatch.
3. **Double resource usage**: The app loads both the sklearn pipeline into memory AND the XGBoost JSON model, consuming roughly double the RAM needed.

## Steps to Reproduce

1. Restart the Streamlit server (or clear the cache with `st.cache_resource.clear()`).
2. Open the app.
3. Observe the ~20-30 second startup delay.
4. Search the entire `application.py` file for usages of the variable `pipe` — it is assigned once and never read.

## Proposed Fix

Remove `train_model()` entirely (or keep it only in the notebook for experimentation). At startup, load the pre-trained `xgb_win_prob_model.json` once using `@st.cache_resource`:

```python
@st.cache_resource
def load_model():
    model = xgb.XGBClassifier()
    model.load_model('xgb_win_prob_model.json')
    return model

loaded_xgb = load_model()  # Cached — loaded once, reused on every prediction
```

Then in the prediction block, replace:
```python
# Remove these two lines:
loaded_xgb = xgb.XGBClassifier()
loaded_xgb.load_model('xgb_win_prob_model.json')
```
with the cached `loaded_xgb` from above.

This reduces cold start time from ~30s to under 1 second and eliminates the dead training code.

## Affected File

`application.py` — `train_model()` function (lines ~1076-1125) and the prediction block (lines ~1522-1524)

## Suggested Labels

`type:bug` `level:advanced` `quality:exceptional`

I would like to work on this issue under GSSoC 2026. Could you please assign it to me? @Arnav-Singh-5080

/assign

### PR DESCRIPTION
**Pull request by anshul23102**
Fixes #99

## Problem

`train_model()` was decorated with `@st.cache_resource`, which means Streamlit caches the *function call itself* and never loads from disk. Every cold start triggers a full CSV read and model fit (20-30 seconds), even when `pipe.pkl` already exists on disk. The cached result also blocks `st.cache_resource.clear()` from working correctly for cache invalidation.

## Changes

- Removed `@st.cache_resource` from `train_model()` so it is a plain offline utility, not a runtime entry point
- Renamed the internal pipeline variable to `trained` to avoid shadowing the module-level `pipe`
- Replaced the `model_name` parameter and `get_model()` helper with a direct `LogisticRegression(max_iter=1000)` call (the multi-model switch was dead code)
- Removed unused imports: `os`, `RandomForestClassifier`, `XGBClassifier`, `logging`, `train_test_split`, `accuracy_score`
- Added `load_pipeline()` decorated with `@st.cache_resource`:
  - Fast path: `joblib.load("pipe.pkl")` (sub-second)
  - Fallback: calls `train_model()`, saves the result to `pipe.pkl`, returns it
- Changed `pipe = train_model()` to `pipe = load_pipeline()`

## Behaviour after this change

| Scenario | Before | After |
|---|---|---|
| `pipe.pkl` present | Ignores it, trains every cold start | Loads from disk in under 1 second |
| `pipe.pkl` absent | Trains from CSV, saves to model_name_model.pkl | Trains from CSV, saves to pipe.pkl |
| Streamlit rerun | Reads from @st.cache_resource memory | Reads from @st.cache_resource memory |

## Test plan

- [ ] Delete `pipe.pkl`, restart the app: confirms fallback trains and writes `pipe.pkl`
- [ ] Restart again with `pipe.pkl` present: confirms sub-second startup with no CSV reads
- [ ] Run Analysis with valid inputs: confirms `pipe.predict_proba` returns correct probabilities
- [ ] Call `st.cache_resource.clear()` and reload: confirms `load_pipeline()` re-reads `pipe.pkl` cleanly

### CODE REVIEW

**Comment by anshul23102** [comment]
Closing this PR as the underlying issue (#99) is already closed and the upstream codebase has since diverged significantly from the base this branch was created on, resulting in unresolvable conflicts.

The current upstream main retains @st.cache_resource on train_model() which serves the same caching purpose for the deployed app. If a load_pipeline() approach is still desired, a fresh branch from current main would be the right way forward.
